### PR TITLE
refactor: remove return bool from addplayer and replaceplayer

### PR DESCRIPTION
**Pull request by uweenukr**
Now if there is an issue it will throw an exception so you know what specifically was wrong.

### CODE REVIEW

**Comment by uweenukr** [comment]
The reason I removed the test is that aside from the AddCharacter() call it is identical to the previous test. The AddCharacter() has no impact on the test. That call was why the test was failing in this PR. Before it would have logged a comment and returned false. Again no impact on the test result. Now its throwing so just remove it.

File: Assets/Mirage/Runtime/ServerObjectManager.cs [hunk 1]
```diff
@@ -236,20 +236,18 @@ public bool AddCharacter(INetworkPlayer player, GameObject character, Guid asset
...
         /// <returns></returns>
-        public bool AddCharacter(INetworkPlayer player, GameObject character)
+        public void AddCharacter(INetworkPlayer player, GameObject character)
         {
             NetworkIdentity identity = character.GetComponent<NetworkIdentity>();
             if (identity is null)
             {
-                logger.Log("AddPlayer: playerGameObject has no NetworkIdentity. Please add a NetworkIdentity to " + character);
-                return false;
+                throw new InvalidOperationException("AddPlayer: playerGameObject has no NetworkIdentity. Please add a NetworkIdentity to " + character);
```
Comment (James-Frowen) [inline Assets/Mirage/Runtime/ServerObjectManager.cs:244]: I think this should be ArgumentException, I think InvalidOperationException is for when SOM would be in an invalid state

File: Assets/Mirage/Runtime/ServerObjectManager.cs [hunk 2]
```diff
@@ -236,20 +236,18 @@ public bool AddCharacter(INetworkPlayer player, GameObject character, Guid asset
...
             }
 
             // cannot have a player object in "Add" version
             if (player.Identity != null)
             {
-                logger.Log("AddPlayer: player object already exists");
-                return false;
+                throw new InvalidOperationException("AddPlayer: player object already exists");
```
Comment (James-Frowen) [inline Assets/Mirage/Runtime/ServerObjectManager.cs:250]: should be ArgumentException as well, see previous comment

File: Assets/Mirage/Runtime/ServerObjectManager.cs [hunk 3]
```diff
@@ -294,19 +291,19 @@ void Respawn(NetworkIdentity identity)
...
 
-        internal bool InternalReplacePlayerForConnection(INetworkPlayer player, INetworkClient client, GameObject character, bool keepAuthority)
+        internal void InternalReplacePlayerForConnection(INetworkPlayer player, INetworkClient client, GameObject character, bool keepAuthority)
         {
             NetworkIdentity identity = character.GetComponent<NetworkIdentity>();
             if (identity is null)
             {
                 logger.LogError("ReplacePlayer: playerGameObject has no NetworkIdentity. Please add a NetworkIdentity to " + character);
-                return false;
+                return;
```
Comment (James-Frowen) [inline Assets/Mirage/Runtime/ServerObjectManager.cs:300]: can this be exception instead of log

File: Assets/Mirage/Runtime/ServerObjectManager.cs [hunk 4]
```diff
@@ -294,19 +291,19 @@ void Respawn(NetworkIdentity identity)
...
             }
 
             if (identity.ConnectionToClient != null && identity.ConnectionToClient != player)
             {
                 logger.LogError("Cannot replace player for connection. New player is already owned by a different connection" + character);
-                return false;
+                return;
```
Comment (James-Frowen) [inline Assets/Mirage/Runtime/ServerObjectManager.cs:306]: can this be exception instead of log

File: Assets/Mirage/Runtime/ServerObjectManager.cs (hunk 2 above)
Comment (uweenukr) [reply Assets/Mirage/Runtime/ServerObjectManager.cs]: fixed

File: Assets/Mirage/Runtime/ServerObjectManager.cs (hunk 3 above)
Comment (uweenukr) [reply Assets/Mirage/Runtime/ServerObjectManager.cs:300]: fixed

File: Assets/Mirage/Runtime/ServerObjectManager.cs (hunk 1 above)
Comment (uweenukr) [reply Assets/Mirage/Runtime/ServerObjectManager.cs]: fixed

File: Assets/Mirage/Runtime/ServerObjectManager.cs (hunk 3 above)
Comment (uweenukr) [reply Assets/Mirage/Runtime/ServerObjectManager.cs:300]: they are both exceptions for AddPlayer. should they both be here?

File: Assets/Tests/Runtime/Host/NetworkIdentityTests.cs [hunk 5]
```diff
@@ -172,7 +172,7 @@ public void RemoveclientAuthorityNotSpawned()
         [Test]
         public void RemoveClientAuthorityOfOwner()
         {
-            serverObjectManager.AddCharacter(server.LocalPlayer, gameObject);
+            serverObjectManager.ReplaceCharacter(server.LocalPlayer, server.LocalClient, gameObject);
```
Comment (James-Frowen) [inline Assets/Tests/Runtime/Host/NetworkIdentityTests.cs:175]: `server.LocalClient` doesn't seem relevant to this test, maybe use default or null instead

File: Assets/Tests/Runtime/ClientServer/ServerObjectManagerTests.cs [hunk 6]
```diff
@@ -249,7 +251,10 @@ public void SpawnObjectExceptionTest()
         [Test]
         public void AddPlayerForConnectionFalseTest()
```
Comment (James-Frowen) [inline Assets/Tests/Runtime/ClientServer/ServerObjectManagerTests.cs]: test names needs to say throws instead of false now

File: Assets/Tests/Runtime/ClientServer/ServerObjectManagerTests.cs [hunk 7]
```diff
@@ -201,8 +201,10 @@ public void AddPlayerForConnectionAssetIdTest()
 
             connectionToClient.Identity = null;
 
-            Assert.That(serverObjectManager.AddCharacter(connectionToClient, playerReplacement, replacementGuid), Is.True);
-        }
+            serverObjectManager.AddCharacter(connectionToClient, playerReplacement, replacementGuid);
+
+            await AsyncUtil.WaitUntilWithTimeout(() => replacementIdentity == connectionToClient.Identity);
```
Comment (James-Frowen) [inline Assets/Tests/Runtime/ClientServer/ServerObjectManagerTests.cs:206]: shouldn't character be set right away? can we just assert it is set instead of await?

File: Assets/Mirage/Runtime/ServerObjectManager.cs (hunk 3 above)
Comment (uweenukr) [reply Assets/Mirage/Runtime/ServerObjectManager.cs]: fixed

File: Assets/Mirage/Runtime/ServerObjectManager.cs (hunk 4 above)
Comment (uweenukr) [reply Assets/Mirage/Runtime/ServerObjectManager.cs]: fixed

File: Assets/Tests/Runtime/ClientServer/ServerObjectManagerTests.cs (hunk 6 above)
Comment (uweenukr) [reply Assets/Tests/Runtime/ClientServer/ServerObjectManagerTests.cs]: fixed

File: Assets/Tests/Runtime/ClientServer/ServerObjectManagerTests.cs (hunk 7 above)
Comment (uweenukr) [reply Assets/Tests/Runtime/ClientServer/ServerObjectManagerTests.cs]: didnt think that would work. turns out it does :). fixed

File: Assets/Tests/Runtime/Host/NetworkIdentityTests.cs (hunk 5 above)
Comment (uweenukr) [reply Assets/Tests/Runtime/Host/NetworkIdentityTests.cs:175]: this will get addressed via: #774

**Review by James-Frowen — APPROVED** [review]

File: Assets/Tests/Runtime/Host/NetworkIdentityTests.cs (hunk 5 above)
Comment (James-Frowen) [inline Assets/Tests/Runtime/Host/NetworkIdentityTests.cs:175]:
```suggestion
            serverObjectManager.ReplaceCharacter(server.LocalPlayer, gameObject);
```

client removed in #774
